### PR TITLE
Added Exploration Mode (Crash Fix)

### PR DESCRIPTION
**Pull request by LordHepipud**
- Exploration mode:
  By default OTClient is drawing all tiles which are send by the server to the minimap view. Some server admins might not want this and will receive a working patch for this. Simply add EXPLORE_MODE to build dependencies to make it work. While still all tiles are received from the server, only the visible ones on the same floor as the player including tiles inside the view range will be drawn. All other tiles will be added to a caching system and drawn later in case the player walks up or down a floor. To prevent the Client from freezing or slowing done after a long play time, the cache will be flushed of all tiles which have been drawn already or which are far out from our view range

Improvements:
- Names of caching variables changed (added m_ before)
- Some code cleanup

Fixes:
- Crash bug while exploring new areas in some cases

### CODE REVIEW

**Comment by diegorodriguesvieira** [comment]
Thanks a lot dude! =)

File: src/client/minimap.cpp [hunk 1]
```diff
@@ -203,12 +209,100 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
         minimapTile.speed = std::min<int>((int)std::ceil(tile->getGroundSpeed() / 10.0f), 255);
     }
 
-    if(minimapTile != MinimapTile()) {
-        MinimapBlock& block = getBlock(pos);
-        Point offsetPos = getBlockOffset(Point(pos.x, pos.y));
-        block.updateTile(pos.x - offsetPos.x, pos.y - offsetPos.y, minimapTile);
-        block.justSaw();
-    }
+#ifndef EXPLORE_MODE
```
Comment (DarkShinz) [inline src/client/minimap.cpp]: This update is an important improvement than just to be defined by the user. If you are uncertain if your code really works, allow me to give you a few tips.

File: src/client/minimap.cpp [hunk 2]
```diff
@@ -203,12 +209,100 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
...
+	//First we need to get the player object
+	LocalPlayerPtr localPlayer = g_game.getLocalPlayer();
+
+	// Do nothing in case the player object is not valid
+	if (localPlayer == nullptr) {
+		return;
+	}
+
+	// Get the current Z-Position and cache the current Player offset based on the player position and tile position
+	short playerPosZ = localPlayer->getPosition().z;
```
Comment (DarkShinz) [inline src/client/minimap.cpp]: I recommend doing it this way: 
`const Position& playerPos = localPlayer->getPosition();`
So you can use x, y, z without calling it again. So whenever you need to do a check you just call 'playerPos.z'

File: src/client/minimap.cpp [hunk 3]
```diff
@@ -203,12 +209,100 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
...
+
+			//Looks like we passed all checks and are capable to view this tile. Lets draw the minimap data
+			it->block->updateTile(it->pos.x - it->offsetPos.x, it->pos.y - it->offsetPos.y, it->minimapTile);
+			it->block->justSaw();
+			it->drawn = true;
+			m_minimapFloorCache[playerPosZ].erase(it);
+			it = cache.begin();
+		}
+	}
+#endif
```
Comment (DarkShinz) [inline src/client/minimap.cpp]: Another tip is to create functions for code that intend to do something different than what the function describes. For example, checking if the MinimapTile should be removed from the cache, should be a function named like `updateMinimapCache()`. This way ppl reading your code would understand what the function does without the need to read comments.

Regarding the code functionality, I want to add my opinion that checks into a cache should be done ONLY when needed. Checking if the cache should be edited every single call of the Minimap::draw function should make significant overhead and impact player's FPS. Take care with those kinds of code.
With that in mind, I believe that a function to check the `updateMinimapCache ` should be called whenever the player position changes or the g_map.tiles updates (taking in consideration you are using the players position to check if minimap tiles are in range. It would be a waste to keep checking it if the position hasn't changed at all. See?).

File: src/client/minimap.cpp [hunk 4]
```diff
@@ -203,12 +209,100 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
...
+		//Create new cache object
+		MinimapCache minimapCache;
+		minimapCache.drawn = false;
+		minimapCache.minimapTile = minimapTile;
+		minimapCache.pos = pos;
+		minimapCache.block = &block;
+		minimapCache.offsetPos = offsetPos;
+
+		//In case the player object is valid and we are capable to view this tile, draw it without caching
+		if (playerPosZ == pos.z && std::abs(localPlayer->getPosition().x - pos.x) <= 9 && std::abs(localPlayer->getPosition().y - pos.y) <= 7) {
```
Comment (DarkShinz) [inline src/client/minimap.cpp]: I see two improvements here:

1.  You need to get the current View Range of the client to use later in position checks. OTC codes should be flexible to this kind of customization, it would look like this:
`AwareRange range = g_map.getAwareRange();`

2. You'd rather use Position::isInRange rather than this check.
It would be something like this (using the previous review about player pos and the previous item):
`if (playerPos.isInRange(pos, range.horizontal(), range.vertical(), 0))`

**Comment by VagnerDomingues** [comment]
interesting

**Comment by spekdrum** [comment]
Hi. I've tested this code and causes a freeze in some cases. For example if you go to a multilevel map zone and go to the top and then go down to the level 6.

**Comment by LordHepipud** [comment]
Thank you for the post. I will have to re-check, but I believe I already run into a similar issue and fixed it locally.
As the feedback was quite low, I expected no one is using it :D

I will keep you updated

**Comment by Klonoa88** [comment]
Same here. This code sometimes freezes when walking between floors below floor 7

**Comment by Klonoa88** [comment]
> 
> 
> Thank you for the post. I will have to re-check, but I believe I already run into a similar issue and fixed it locally.
> As the feedback was quite low, I expected no one is using it :D
> 
> I will keep you updated

Same here. This code sometimes freezes when walking between floors below floor 7

**Comment by LordHepipud** [comment]
I just made a commit with some changes. Mainly the issue was caused by

```c++
if (it->drawn == true) {
    m_minimapFloorCache[playerPosZ].erase(it);
    it = cache.begin();
    continue;
}
```

which is not handled differently. Can you please test if the new code works/compiles? I have an issue with the old version of my source and sadly cant compile it for testing if I missed anything.
This part is working on my server without any freezes and crashes for a very long time.

File: modules/corelib/ui/uiscrollbar.lua [hunk 5]
```diff
@@ -28,7 +28,7 @@ local function calcValues(self)
     proportion = math.min(math.max(self.step, 1), range)/range
   end
 
-  local px = math.max(proportion * pxrange, 10)
+  local px = math.max(proportion * pxrange, 6)
```
Comment (slavidodo) [inline modules/corelib/ui/uiscrollbar.lua:31]: is this relevant to the commit?

File: src/client/minimap.cpp [hunk 6]
```diff
@@ -24,6 +24,9 @@
 #include "minimap.h"
 #include "tile.h"
 
+//PhiadariaSoft Mod: Minimap drawing fix to prevent drwaing multiple floors
```
Comment (slavidodo) [inline src/client/minimap.cpp:27]: remove that line or add it to where it fits

File: src/client/minimap.cpp [hunk 7]
```diff
@@ -96,6 +99,9 @@ void Minimap::clean()
 {
     for(int i=0;i<=Otc::MAX_Z;++i)
         m_tileBlocks[i].clear();
+        
+    m_minimapFloorCache.clear();
```
Comment (slavidodo) [inline src/client/minimap.cpp:103]: indentation

File: src/client/minimap.cpp [hunk 8]
```diff
@@ -203,12 +209,91 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
         minimapTile.speed = std::min<int>((int)std::ceil(tile->getGroundSpeed() / 10.0f), 255);
     }
 
-    if(minimapTile != MinimapTile()) {
+#ifndef EXPLORE_MODE
+    if (minimapTile != MinimapTile()) {
```
Comment (slavidodo) [inline src/client/minimap.cpp:213]: follow the style here to make it compatible with all the code!
``if(x)``

File: src/client/minimap.h [hunk 9]
```diff
@@ -73,6 +73,15 @@ class MinimapBlock
     stdext::boolean<false> m_wasSeen;
 };
 
+//EXPLORATION MODE
```
Comment (slavidodo) [inline src/client/minimap.h:76]: use ```#ifdef EXPLORE_MODE``

File: src/client/minimap.h [hunk 10]
```diff
@@ -107,6 +116,14 @@ class Minimap
                                                                   (index / (65536 / MMBLOCK_SIZE))*MMBLOCK_SIZE, z); }
     uint getBlockIndex(const Position& pos) { return ((pos.y / MMBLOCK_SIZE) * (65536 / MMBLOCK_SIZE)) + (pos.x / MMBLOCK_SIZE); }
     std::unordered_map<uint, MinimapBlock> m_tileBlocks[Otc::MAX_Z+1];
+    
+    //EXPLORATION MODE
```
Comment (slavidodo) [inline src/client/minimap.h:120]: use ```#ifdef EXPLORE_MODE``

File: src/client/minimap.cpp [hunk 11]
```diff
@@ -203,12 +209,91 @@ void Minimap::updateTile(const Position& pos, const TilePtr& tile)
...
+            }
+
+            // Looks like we passed all checks and are capable to view this tile. Lets draw the minimap data
+            it->block->updateTile(it->pos.x - it->offsetPos.x, it->pos.y - it->offsetPos.y, it->minimapTile);
+            it->block->justSaw();
+            m_minimapFloorCache[playerPosZ].erase(it);
+            it = cache.begin();
+        }
+    }
+#endif
```
Comment (slavidodo) [inline src/client/minimap.cpp:296]: follow the style we're using in the code! single-line if statement, ...

File: modules/corelib/ui/uiscrollbar.lua (hunk 5 above)
Comment (LordHepipud) [reply modules/corelib/ui/uiscrollbar.lua:31]: No this shouldn't be a part of it. No idea how this was added Oo

File: src/client/minimap.cpp (hunk 11 above)
Comment (LordHepipud) [reply src/client/minimap.cpp:296]: What? Oo
Where is the problem here? Oo

Comment (slavidodo) [reply src/client/minimap.cpp:296]:
```cpp
if(x)
    DoSomething();

for(x){
```

File: src/client/minimap.cpp (hunk 8 above)
Comment (slavidodo) [reply src/client/minimap.cpp:213]: @LordHepipud just like here :)

**Comment by diath** [comment]
A much simpler approach would be to simply not update the minimap when the z coordinate is different, I don't see much reason to cache the previously seen tiles, if the player wasn't to that place yet then it makes sense that they should discover that area on their own once their get to the destination floor:
```diff --git a/src/client/map.cpp b/src/client/map.cpp
index 3f768c257234b9dc79784f52460c170b34c46543..6f7db24229e9e9a54f5abaf14cf6860afd40d08c 100644
--- a/src/client/map.cpp
+++ b/src/client/map.cpp
@@ -44,7 +44,11 @@ void Map::notificateTileUpdate(const Position& pos)
 
     for(const MapViewPtr& mapView : m_mapViews)
         mapView->onTileUpdate(pos);
-    g_minimap.updateTile(pos, getTile(pos));
+
+    LocalPlayerPtr localPlayer = g_game.getLocalPlayer();
+    if (localPlayer && localPlayer->getPosition().z == pos.z) {
+        g_minimap.updateTile(pos, getTile(pos));
+    }
 }
 
 void Map::clean()
```

Also it would be better if this was behind a runtime flag, not a compile time flag.
